File: Source/Forums/ForumListViewController.m
//  ForumListViewController.m
//
//  Copyright 2010 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

#import "ForumListViewController.h"
#import <AFNetworking/AFNetworking.h>
#import "AwfulForumsClient.h"
#import "AwfulForumTreeDataSource.h"
#import "AwfulFrameworkCategories.h"
#import "AwfulRefreshMinder.h"
#import "AwfulSettings.h"
#import "ThreadListViewController.h"
#import "Awful-Swift.h"

@interface ForumListViewController () <AwfulForumTreeDataSourceDelegate>

@property (nonatomic) NSMutableArray *favoriteForums;
@property (nonatomic) BOOL userDrivenChange;

@end

@implementation ForumListViewController
{
    AwfulForumTreeDataSource *_treeDataSource;
    BOOL _observingReachability;
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (instancetype)initWithManagedObjectContext:(NSManagedObjectContext *)managedObjectContext
{
    if ((self = [[UIStoryboard storyboardWithName:@"ForumList" bundle:nil] instantiateInitialViewController])) {
        _managedObjectContext = managedObjectContext;
        _favoriteForums = [[self fetchFavoriteForumsWithIDsFromSettings] mutableCopy];
        
        self.title = @"Forums";
        self.navigationItem.backBarButtonItem = [UIBarButtonItem awful_emptyBackBarButtonItem];
        self.tabBarItem.accessibilityLabel = @"Forums list";
        self.tabBarItem.image = [UIImage imageNamed:@"list_icon"];
        [self showOrHideEditButton];
        
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(settingsDidChange:) name:AwfulSettingsDidChangeNotification object:nil];
    }
    return self;
}

- (NSArray *)fetchFavoriteForumsWithIDsFromSettings
{
    NSArray *forumIDs = [AwfulSettings sharedSettings].favoriteForums;
    if (forumIDs.count == 0) {
        return @[];
    }
    NSArray *favoriteForums = [Forum fetchAllInManagedObjectContext:self.managedObjectContext
                                            matchingPredicateFormat:@"forumID IN %@", forumIDs];
    return [favoriteForums sortedArrayUsingComparator:^(Forum *a, Forum *b) {
        return [@([forumIDs indexOfObject:a.forumID]) compare:@([forumIDs indexOfObject:b.forumID])];
    }];
}

- (void)settingsDidChange:(NSNotification *)note
{
    if (self.userDrivenChange) return;
    
    NSString *changedSetting = note.userInfo[AwfulSettingsDidChangeSettingKey];
    if ([changedSetting isEqualToString:AwfulSettingsKeys.favoriteForums]) {
        [self.favoriteForums setArray:[self fetchFavoriteForumsWithIDsFromSettings]];
        [self showOrHideEditButton];
        [self.tableView reloadData];
    }
}

- (void)loadView
{
    [super loadView];
    [self.tableView registerClass:[UITableViewHeaderFooterView class] forHeaderFooterViewReuseIdentifier:HeaderIdentifier];
}

static NSString * const HeaderIdentifier = @"Header";

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.tableView.estimatedRowHeight = 44;
    self.tableView.backgroundView = nil;
    [self.tableView awful_unstickSectionHeaders];
    _treeDataSource = [[AwfulForumTreeDataSource alloc] initWithTableView:self.tableView reuseIdentifier:@"Forum"];
    _treeDataSource.topDataSource = self;
    _treeDataSource.managedObjectContext = self.managedObjectContext;
    _treeDataSource.delegate = self;
}

- (void)showOrHideEditButton
{
    UIBarButtonItem *item = self.favoriteForums.count > 0 ? self.editButtonItem : nil;
    [self.navigationItem setRightBarButtonItem:item animated:YES];
}

- (void)toggleFavorite:(UIButton *)button
{
    // Figure out which forum is represented by the tapped button's cell.
    Forum *forum;
    UIView *cell = button.superview;
    while (cell && ![cell isKindOfClass:[UITableViewCell class]]) {
        cell = cell.superview;
    }
    if (!cell) return;
    NSIndexPath *indexPath = [self.tableView indexPathForCell:(UITableViewCell *)cell];
    if (self.favoriteForums.count > 0 && indexPath.section == 0) {
        forum = self.favoriteForums[indexPath.row];
    } else {
        forum = [_treeDataSource forumAtIndexPath:indexPath];
    }
    BOOL isFavorite = [self.favoriteForums containsObject:forum];
    
    [self.tableView beginUpdates];
    
    [_treeDataSource reloadRowWithForum:forum];
    
    if (isFavorite) {
        NSInteger row = [self.favoriteForums indexOfObject:forum];
        [self.favoriteForums removeObjectAtIndex:row];
        if (self.favoriteForums.count == 0) {
            [self.tableView deleteSections:[NSIndexSet indexSetWithIndex:0] withRowAnimation:UITableViewRowAnimationTop];
            [self showOrHideEditButton];
            self.editing = NO;
        } else {
            NSIndexPath *oldRowPath = [NSIndexPath indexPathForRow:row inSection:0];
            [self.tableView deleteRowsAtIndexPaths:@[ oldRowPath ] withRowAnimation:UITableViewRowAnimationTop];
        }
    } else {
        [self.favoriteForums addObject:forum];
        if (self.favoriteForums.count == 1) {
            [self.tableView insertSections:[NSIndexSet indexSetWithIndex:0] withRowAnimation:UITableViewRowAnimationTop];
            [self showOrHideEditButton];
        } else {
            NSIndexPath *newRowPath = [NSIndexPath indexPathForRow:(self.favoriteForums.count - 1) inSection:0];
            [self.tableView insertRowsAtIndexPaths:@[ newRowPath ] withRowAnimation:UITableViewRowAnimationAutomatic];
        }
    }
    
    [self.tableView endUpdates];
    
    [self saveFavoriteForumsToSettings];
}

- (void)toggleExpanded:(UIButton *)button
{
    button.selected = !button.selected;
    UpdateDisclosureButtonAccessibilityLabel(button);
    UIView *cell = button.superview;
    while (cell && ![cell isKindOfClass:[UITableViewCell class]]) {
        cell = cell.superview;
    }
    if (!cell) return;
    NSIndexPath *indexPath = [self.tableView indexPathForCell:(UITableViewCell *)cell];
    Forum *forum = [_treeDataSource forumAtIndexPath:indexPath];
    [_treeDataSource setForum:forum childrenExpanded:button.selected];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    _treeDataSource.updatesTableView = YES;
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    [self refreshIfNecessary];
}

- (void)refreshIfNecessary
{
    if (![AwfulForumsClient client].loggedIn) return;
    
    if ([[AwfulRefreshMinder minder] shouldRefreshForumList] || self.tableView.numberOfSections < 2 ||
        [Forum anyInManagedObjectContext:self.managedObjectContext matchingPredicateFormat:@"index = -1"]) {
        if ([AwfulForumsClient client].reachable) {
            [self refresh];
        } else {
            [self refreshOnceServerIsReachable];
        }
    }
}

- (void)refresh
{
    [[AwfulForumsClient client] taxonomizeForumsAndThen:^(NSError *error, NSArray *categories) {
        if (!error) {
            [[AwfulRefreshMinder minder] didFinishRefreshingForumList];
        }
    }];
}

- (void)refreshOnceServerIsReachable
{
    if (_observingReachability) return;
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(reachabilityChanged:)
                                                 name:AFNetworkingReachabilityDidChangeNotification
                                               object:nil];
    _observingReachability = YES;

}

- (void)reachabilityChanged:(NSNotification *)note
{
    [self stopObservingReachability];
    [self refresh];
}

- (void)stopObservingReachability
{
    if (!_observingReachability) return;
    [[NSNotificationCenter defaultCenter] removeObserver:self name:AFNetworkingReachabilityDidChangeNotification object:nil];
    _observingReachability = NO;
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    _treeDataSource.updatesTableView = NO;
}

- (void)showForum:(Forum *)forum animated:(BOOL)animated
{
    ThreadListViewController *threadList = [[ThreadListViewController alloc] initWithForum:forum];
    threadList.restorationClass = threadList.class;
    threadList.restorationIdentifier = @"Thread";
    [self.navigationController pushViewController:threadList animated:animated];
}

- (void)saveFavoriteForumsToSettings
{
    self.userDrivenChange = YES;
    [AwfulSettings sharedSettings].favoriteForums = [self.favoriteForums valueForKey:@"forumID"];
    self.userDrivenChange = NO;
}

static void UpdateDisclosureButtonAccessibilityLabel(UIButton *button) {
    if (button.selected) {
        button.accessibilityLabel = @"Hide subforums";
    } else {
        button.accessibilityLabel = @"Expand subforums";
    }
}

#pragma mark - UITableViewDataSource

// AwfulForumsListController works in concert with an AwfulForumTreeDataSource, setting itself as the tree data source's topDataSource. As such, we only need to handle the UITableViewDataSource methods for the section we control: section 0, "Favorites".

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    if (self.favoriteForums.count > 0) {
        return 1;
    }
    return 0;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.favoriteForums.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    ForumCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Favorite" forIndexPath:indexPath];
    Forum *forum = self.favoriteForums[indexPath.row];
    cell.nameLabel.text = forum.name;
    ThemeCell(self.theme, cell);
    return cell;
}

static void ThemeCell(AwfulTheme *theme, ForumCell *cell)
{
    cell.backgroundColor = theme[@"listBackgroundColor"];
	cell.nameLabel.textColor = theme[@"listTextColor"];
    cell.separator.backgroundColor = theme[@"listSeparatorColor"];
    UIView *selectedBackgroundView = [UIView new];
    selectedBackgroundView.backgroundColor = theme[@"listSelectedBackgroundColor"];
    cell.selectedBackgroundView = selectedBackgroundView;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)favoriteIndexPath
{
    Forum *forum = self.favoriteForums[favoriteIndexPath.row];
    
    // Let's delete the favorite row and update the forum's proper row in one fell swoop. To do so, we need the forum's index path before any potential deletion of the favorites section (in model or in view).
    NSIndexPath *forumIndexPath = [_treeDataSource indexPathForForum:forum];
    
    // Now it's safe to update the array of favorite forums.
    [self.favoriteForums removeObjectAtIndex:favoriteIndexPath.row];
    
    [tableView beginUpdates];
    
    if (self.favoriteForums.count == 0) {
        [tableView deleteSections:[NSIndexSet indexSetWithIndex:0] withRowAnimation:UITableViewRowAnimationTop];
        [self showOrHideEditButton];
        self.editing = NO;
    } else {
        [tableView deleteRowsAtIndexPaths:@[ favoriteIndexPath ] withRowAnimation:UITableViewRowAnimationAutomatic];
    }
    if (forumIndexPath) {
        [tableView reloadRowsAtIndexPaths:@[ forumIndexPath ] withRowAnimation:UITableViewRowAnimationNone];
    }
    
    [tableView endUpdates];
    
    [self saveFavoriteForumsToSettings];
}

- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath
{
    return YES;
}

- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)sourceIndexPath toIndexPath:(NSIndexPath *)destinationIndexPath
{
    id forum = self.favoriteForums[sourceIndexPath.row];
    [self.favoriteForums removeObjectAtIndex:sourceIndexPath.row];
    [self.favoriteForums insertObject:forum atIndex:destinationIndexPath.row];
    [self saveFavoriteForumsToSettings];
}

#pragma mark - UITableViewDelegate

- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section
{
    UITableViewHeaderFooterView *header = [tableView dequeueReusableHeaderFooterViewWithIdentifier:HeaderIdentifier];
    header.textLabel.textColor = self.theme[@"listHeaderTextColor"];
    header.contentView.backgroundColor = self.theme[@"listHeaderBackgroundColor"];
    if (self.favoriteForums.count > 0 && section == 0) {
        header.textLabel.text = @"Favorites";
    } else {
        header.textLabel.text = [_treeDataSource categoryNameAtIndex:section];
    }
    return header;
}

- (void)tableView:(UITableView *)tableView willDisplayHeaderView:(UITableViewHeaderFooterView *)header forSection:(NSInteger)section
{
    // For some unknown reason, setting this font apparently needs to happen here.
    UIFontDescriptor *descriptor = [UIFontDescriptor preferredFontDescriptorWithTextStyle:UIFontTextStyleBody];
    header.textLabel.font = [UIFont fontWithDescriptor:descriptor size:MIN(descriptor.pointSize, 23)];
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section
{
    return 36;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    Forum *forum;
    if (self.favoriteForums.count > 0 && indexPath.section == 0) {
        forum = self.favoriteForums[indexPath.row];
    } else {
        forum = [_treeDataSource forumAtIndexPath:indexPath];
    }
    [self showForum:forum animated:YES];
}

- (BOOL)tableView:(UITableView *)tableView shouldIndentWhileEditingRowAtIndexPath:(NSIndexPath *)indexPath
{
    return self.favoriteForums.count > 0 && indexPath.section == 0;
}

- (NSIndexPath *)tableView:(UITableView *)tableView targetIndexPathForMoveFromRowAtIndexPath:(NSIndexPath *)sourceIndexPath toProposedIndexPath:(NSIndexPath *)proposedDestinationIndexPath
{
    if (proposedDestinationIndexPath.section == 0) return proposedDestinationIndexPath;
    return [NSIndexPath indexPathForRow:[tableView numberOfRowsInSection:0] - 1 inSection:0];
}

#pragma mark - AwfulForumTreeDataSourceDelegate

- (void)configureCell:(ForumListCell *)cell withForum:(Forum *)forum
{
    BOOL hasSubforums = forum.childForums.count > 0;
    cell.disclosureButton.hidden = !hasSubforums;
    if (hasSubforums) {
        [cell.disclosureButton addTarget:self action:@selector(toggleExpanded:) forControlEvents:UIControlEventTouchUpInside];
        cell.disclosureButton.selected = [_treeDataSource forumChildrenExpanded:forum];
    }
    UpdateDisclosureButtonAccessibilityLabel(cell.disclosureButton);
    cell.nameLabel.text = forum.name;
    BOOL isFavorite = [self.favoriteForums containsObject:forum];
    cell.favoriteButton.hidden = isFavorite;
    if (!isFavorite) {
        [cell.favoriteButton addTarget:self action:@selector(toggleFavorite:) forControlEvents:UIControlEventTouchUpInside];
    }
    
    NSMutableString *accessibilityLabel = [cell.nameLabel.text mutableCopy];
    if (hasSubforums) {
        [accessibilityLabel appendFormat:@". %lu subforum%@", (unsigned long)forum.childForums.count, forum.childForums.count == 1 ? @"" : @"s"];
    }
    cell.accessibilityLabel = accessibilityLabel;
    
    NSInteger subforumLevel = 0;
    Forum *currentForum = forum.parentForum;
    while (currentForum) {
        subforumLevel++;
        currentForum = currentForum.parentForum;
    }
    cell.subforumDepth = subforumLevel;
    ThemeCell(self.theme, cell);
}

@end

File: Source/Forums/ForumListViewController.swift
//  ForumListViewController.swift
//
//  Copyright 2014 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

@objc(ForumListViewController)
class ForumListViewController: AwfulTableViewController {
    let managedObjectContext: NSManagedObjectContext
    private let dataSource: DataSource
    private let favoriteDataSource: ForumFavoriteDataSource
    private let treeDataSource: ForumTreeDataSource
    private let contextObserver: ForumContextObserver!

    required init(coder: NSCoder) {
        managedObjectContext = AwfulAppDelegate.instance().managedObjectContext
        favoriteDataSource = ForumFavoriteDataSource(managedObjectContext: managedObjectContext)
        treeDataSource = ForumTreeDataSource(managedObjectContext: managedObjectContext)
        dataSource = CompoundDataSource(favoriteDataSource, treeDataSource)
        super.init(coder: coder)
        dataSource.delegate = self
        
        // Since the data sources work on ForumMetadata entities, it won't pick up any changes to the names of the forums. We'll handle that here.
        contextObserver = ForumContextObserver(managedObjectContext: managedObjectContext) { [unowned self] forums in
            if self.editing {
                // Don't do anything if we're rearranging favorites. It's the same issue mentioned by the comment in dataSource(_:performBatchUpdates:completion:).
                return
            }
            self.tableView.beginUpdates()
            for forum in forums {
                let indexPaths = self.dataSource.indexPathsForItem(forum.metadata)
                self.dataSource(self.dataSource, didRefreshItemsAtIndexPaths: indexPaths)
            }
            self.tableView.endUpdates()
        }
        
        navigationItem.backBarButtonItem = UIBarButtonItem.awful_emptyBackBarButtonItem()
    }
    
    class func newFromStoryboard() -> ForumListViewController {
        return UIStoryboard(name: "ForumList", bundle: nil).instantiateInitialViewController() as ForumListViewController
    }
    
    private class ForumContextObserver: NSObject {
        let managedObjectContext: NSManagedObjectContext
        let entity: NSEntityDescription
        let changeBlock: [Forum] -> Void
        
        init(managedObjectContext context: NSManagedObjectContext, changeBlock: [Forum] -> Void) {
            managedObjectContext = context
            entity = NSEntityDescription.entityForName(Forum.entityName(), inManagedObjectContext: context)!
            self.changeBlock = changeBlock
            super.init()
            NSNotificationCenter.defaultCenter().addObserver(self, selector: "objectsDidChange:", name: NSManagedObjectContextObjectsDidChangeNotification, object: context)
        }
        
        deinit {
            NSNotificationCenter.defaultCenter().removeObserver(self)
        }
        
        @objc private func objectsDidChange(notification: NSNotification) {
            let userInfo = notification.userInfo as [String:AnyObject]
            let changedObjects = NSMutableSet()
            for key in [NSUpdatedObjectsKey, NSRefreshedObjectsKey] {
                if let updated = userInfo[key] as NSSet? {
                    changedObjects.unionSet(updated)
                }
            }
            let forums = filter(changedObjects) { ($0 as NSManagedObject).entity == self.entity } as [Forum]
            if !forums.isEmpty {
                changeBlock(forums)
            }
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Can only get down to 1 in IB. Pretend that means 0.
        if tableView.sectionHeaderHeight == 1 {
            tableView.sectionHeaderHeight = 0
        }
        if tableView.sectionFooterHeight == 1 {
            tableView.sectionFooterHeight = 0
        }

        tableView.registerClass(UITableViewHeaderFooterView.self, forHeaderFooterViewReuseIdentifier: headerIdentifier)
        tableView.dataSource = dataSource
        
        updateEditButtonPresence(animated: false)
    }
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        refreshIfNecessary()
    }
    
    private func refreshIfNecessary() {
        if AwfulRefreshMinder.sharedMinder().shouldRefreshForumList() || dataSource.numberOfSections == 0 {
            refresh()
        }
    }
    
    @IBAction private func refresh() {
        AwfulForumsClient.sharedClient().taxonomizeForumsAndThen { error, forums in
            if error == nil {
                AwfulRefreshMinder.sharedMinder().didFinishRefreshingForumList()
                self.migrateFavoriteForumsFromSettings()
            }
            self.refreshControl?.endRefreshing()
        }
    }
    
    private func migrateFavoriteForumsFromSettings() {
        // In Awful 3.2 favorite forums moved from AwfulSettings (i.e. NSUserDefaults) to the ForumMetadata entity in Core Data.
        if let forumIDs = AwfulSettings.sharedSettings().favoriteForums as [String]? {
            AwfulSettings.sharedSettings().favoriteForums = nil
            let metadatas = ForumMetadata.metadataForForumsWithIDs(forumIDs, inManagedObjectContext: managedObjectContext)
            for (i, metadata) in enumerate(metadatas) {
                metadata.favoriteIndex = Int32(i)
                metadata.favorite = true
            }
            var error: NSError?
            if !managedObjectContext.save(&error) {
                fatalError("error saving: \(error)")
            }
        }
    }
    
    @IBAction private func didToggleShowingChildren(sender: UIButton) {
        sender.selected = !sender.selected
        let cell: UITableViewCell = sender.nearestSuperviewOfDeclaredType()
        let indexPath = tableView.indexPathForCell(cell)!
        let metadata = dataSource.itemAtIndexPath(indexPath) as ForumMetadata
        metadata.showsChildrenInForumList = sender.selected
        metadata.updateSubtreeVisibility()
    }
    
    @IBAction private func didTapFavoriteStar(sender: UIButton) {
        let cell: UITableViewCell = sender.nearestSuperviewOfDeclaredType()
        let indexPath = tableView.indexPathForCell(cell)!
        let metadata = dataSource.itemAtIndexPath(indexPath) as ForumMetadata
        metadata.favoriteIndex = Int32(favoriteDataSource.fetchedResultsController.fetchedObjects?.count ?? 0)
        metadata.favorite = true
        
        // Trigger a refresh for just this cell.
        metadata.willChangeValueForKey("visibleInForumList")
        metadata.didChangeValueForKey("visibleInForumList")
    }

    func openForum(forum: Forum, animated: Bool) {
        let threadList = ThreadListViewController(forum: forum)
        threadList.restorationClass = ThreadListViewController.self
        threadList.restorationIdentifier = "Thread"
        navigationController?.pushViewController(threadList, animated: animated)
    }
    
    override func setEditing(editing: Bool, animated: Bool) {
        // Since animating table changes while we rearrange favorites causes problems, we'll just do a full reload when we stop rearranging favorites. If we're animating the end of editing, we'll do it after the animation. Otherwise we'll just do it at the end.
        let reloadBlock = { self.tableView.reloadData() }
        if animated && !editing {
            CATransaction.begin()
            CATransaction.setCompletionBlock(reloadBlock)
        }
        
        super.setEditing(editing, animated: animated)
        updateEditButtonPresence(animated: animated)
        
        if !editing {
            if animated {
                CATransaction.commit()
            } else {
                reloadBlock()
            }
        }
    }
    
    private func updateEditButtonPresence(#animated: Bool) {
        navigationItem.setRightBarButtonItem(favoriteDataSource.numberOfSections > 0 ? editButtonItem() : nil, animated: animated)
    }
}

extension UIView {
    func nearestSuperviewOfDeclaredType<T: UIView>() -> T {
        var currentView: UIView! = self
        while currentView != nil {
            // `if currentView is T` doesn't work here and I'm not sure why (T seems to get stuck as UIView somehow and self.superview is invariably returned).
            if currentView.isKindOfClass(T) {
                return currentView as T
            } else {
                currentView = currentView.superview
            }
        }
        fatalError("could not find superview of type \(T.self)")
    }
}

extension ForumMetadata {
    func updateSubtreeVisibility() {
        let childMetadatas = map(forum.childForums) { ($0 as Forum).metadata }
        for child in childMetadatas {
            if showsChildrenInForumList {
                child.visibleInForumList = visibleInForumList
            } else {
                child.visibleInForumList = false
            }
            child.updateSubtreeVisibility()
        }
    }
}

extension ForumListViewController: UITableViewDelegate {
    override func tableView(tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        let header = tableView.dequeueReusableHeaderFooterViewWithIdentifier(headerIdentifier) as UITableViewHeaderFooterView
        header.textLabel.text = dataSource.tableView?(tableView, titleForHeaderInSection: section)
        return header
    }
    
    override func tableView(tableView: UITableView, willDisplayHeaderView view: UIView, forSection section: Int) {
        if let header = view as? UITableViewHeaderFooterView {
            header.textLabel.textColor = theme["listHeaderTextColor"] as UIColor?
            header.contentView.backgroundColor = theme["listHeaderBackgroundColor"] as UIColor?
            header.textLabel.font = UIFont.preferredFontForTextStyle(UIFontTextStyleBody)
        }
    }
    
    override func tableView(tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return UIFont.preferredFontForTextStyle(UIFontTextStyleBody).pointSize * 2
    }
    
    override func tableView(tableView: UITableView, estimatedHeightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        // This method is implemented not for scrolling efficiency but because a missing implementation wreaks havoc on cell heights when expanding or collapsing a forum's subforums.
        return ForumCell.minimumHeight()
    }
    
    override func tableView(tableView: UITableView, willDisplayCell cell: UITableViewCell, forRowAtIndexPath indexPath: NSIndexPath) {
        if let cell = cell as? ForumCell {
            cell.nameLabel.textColor = theme["listTextColor"] as UIColor?
            cell.backgroundColor = theme["listBackgroundColor"] as UIColor?
            cell.selectedBackgroundColor = theme["listSelectedBackgroundColor"] as UIColor?
            if indexPath.row + 1 == tableView.numberOfRowsInSection(indexPath.section) {
                cell.separator.backgroundColor = cell.backgroundColor
            } else {
                cell.separator.backgroundColor = theme["listSeparatorColor"] as UIColor?
            }
        }
    }
    
    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        let metadata = dataSource.itemAtIndexPath(indexPath) as ForumMetadata
        openForum(metadata.forum, animated: true)
    }
    
    override func tableView(tableView: UITableView, titleForDeleteConfirmationButtonForRowAtIndexPath indexPath: NSIndexPath) -> String! {
        return dataSource.tableView?(tableView, titleForDeleteConfirmationButtonForRowAtIndexPath: indexPath)
    }
    
    override func tableView(tableView: UITableView, targetIndexPathForMoveFromRowAtIndexPath sourceIndexPath: NSIndexPath, toProposedIndexPath proposedDestinationIndexPath: NSIndexPath) -> NSIndexPath {
        return dataSource.tableView?(tableView, targetIndexPathForMoveFromRowAtIndexPath: sourceIndexPath, toProposedIndexPath: proposedDestinationIndexPath) ?? proposedDestinationIndexPath
    }
}

private let headerIdentifier = "Header"

extension ForumListViewController: DataSourceDelegate {
    func dataSource(dataSource: DataSource, didInsertItemsAtIndexPaths indexPaths: [NSIndexPath]) {
        tableView.insertRowsAtIndexPaths(indexPaths, withRowAnimation: .Top)
    }
    
    func dataSource(dataSource: DataSource, didRemoveItemsAtIndexPaths indexPaths: [NSIndexPath]) {
        tableView.deleteRowsAtIndexPaths(indexPaths, withRowAnimation: .Automatic)
    }
    
    func dataSource(dataSource: DataSource, didRefreshItemsAtIndexPaths indexPaths: [NSIndexPath]) {
        tableView.reloadRowsAtIndexPaths(indexPaths, withRowAnimation: .None)
    }
    
    func dataSource(dataSource: DataSource, didMoveItemAtIndexPath fromIndexPath: NSIndexPath, toIndexPath: NSIndexPath) {
        tableView.moveRowAtIndexPath(fromIndexPath, toIndexPath: toIndexPath)
    }

    func dataSource(dataSource: DataSource, didInsertSections sections: NSIndexSet) {
        tableView.insertSections(sections, withRowAnimation: .Top)
    }
    
    func dataSource(dataSource: DataSource, didRemoveSections sections: NSIndexSet) {
        tableView.deleteSections(sections, withRowAnimation: .Automatic)
    }
    
    func dataSource(dataSource: DataSource, didRefreshSections sections: NSIndexSet) {
        tableView.reloadSections(sections, withRowAnimation: .None)
    }
    
    func dataSource(dataSource: DataSource, didMoveSection fromSection: Int, toSection: Int) {
        tableView.moveSection(fromSection, toSection: toSection)
    }
    
    func dataSourceDidReloadData(dataSource: DataSource) {
        tableView.reloadData()
    }
    
    func dataSource(dataSource: DataSource, performBatchUpdates updates: () -> Void, completion: (() -> Void)?) {
        // Moving favorites around triggers updates to the tree part of the table. Unfortunately, if we perform those updates while the moved rows are animating, the rows involved in the move get horribly deformed. This is a pretty stupid workaround, but here we are: don't bother updating the table if we're editing. We'll reload the table once we're done rearranging favorites.
        if !editing {
            tableView.beginUpdates()
            updates()
            tableView.endUpdates()
        }
        completion?()
        
        if favoriteDataSource.numberOfSections == 0 {
            setEditing(false, animated: true)
        } else {
            updateEditButtonPresence(animated: true)
        }
    }
}
